src/react/reconcilation.js

```diff
@@ -678,8 +679,20 @@ export class Reconciler {
   ): Value {
     // create a new simple instance of this React class component
     let instance = createClassInstanceForFirstRenderOnly(this.realm, componentType, props, context, evaluatedNode);
+    let getDerivedStateFromProps = Get(this.realm, componentType, "getDerivedStateFromProps");
+
+    if (getDerivedStateFromProps instanceof ECMAScriptSourceFunctionValue && getDerivedStateFromProps.$Call) {
+      applyGetDerivedStateFromProps(this.realm, getDerivedStateFromProps, instance, props);
+    }
     // get the "componentWillMount" and "render" methods off the instance
     let componentWillMount = Get(this.realm, instance, "componentWillMount");
+
+    if (componentWillMount === this.realm.intrinsics.undefined) {
+      // try the unsafe version, added in React 16.3
+      // at some point we only want to remove all support for these unsafe
+      // lifecycle events
+      componentWillMount = Get(this.realm, instance, "UNSAFE_componentWillMount");
+    }
```

Comment:
I think this logic is wrong and should mirror what React does more closely.

Specifically:

* If static `getDerivedStateFromProps` or instance `getSnapshotBeforeUpdate` exist, we treat component as having a “modern” lifecycle. In that case neither `componentWillMount` nor `UNSAFE_componentWillMount` gets called ([ref](https://github.com/facebook/react/blob/915bb5321a8db3435eb36ef1cf9414c15333b447/packages/react-reconciler/src/ReactFiberClassComponent.js#L720-L725)).

* If neither of the modern methods exist, then we first call `componentWillMount` if it exists, then call `UNSAFE_componentWillMount` if it exists. Even if `componentWillMount` exists, it should *not* prevent `UNSAFE_componentWillMount` from getting called. ([ref](https://github.com/facebook/react/blob/915bb5321a8db3435eb36ef1cf9414c15333b447/packages/react-reconciler/src/ReactFiberClassComponent.js#L581-L586))

Reply:
Made the relevant changes.